src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1008,11 +1008,29 @@ int32_t SystemNative_PosixFAdvise(intptr_t fd, int64_t offset, int64_t length, i
 #endif
 }
 
-int32_t SystemNative_PosixFAllocate(intptr_t fd, int64_t offset, int64_t length)
+int32_t SystemNative_PosixFAllocate(intptr_t fd, int64_t offset, int64_t length, int32_t mode)
 {
     assert_msg(offset == 0, "Invalid offset value", (int)offset);
 
     int fileDescriptor = ToFileDescriptor(fd);
+#if !HAVE_POSIX_FALLOCATE64 && !HAVE_POSIX_FALLOCATE
+    // posix_fallocate does not allow for shrinking file
+    // when it's not available, we need to ensure that the file won't be shrinked
+    if (mode == 4) // OpenOrCreate
```

Comment:
Why only OpenOrCreate?  Does preallocationSize not apply when Open'ing an existing file?  And if not, why would we preallocate the size when opening an existing file with OpenOrCreate but not when opening an existing file with Open?

Reply:
That is a great point. I've changed the implementation to apply only to new files, no matter if `FileMode.OpenOrCreate` or `FileMode.Open` was used. It has also simplified the implementation